### PR TITLE
Improve containers API

### PR DESCRIPTION
**Pull request by pcarranzav**
Add default binds to containers created with API, add /v1/containers/update endpoint, and add a /data bind with an internal id.

connects to #215 

Containers now respect the settings that Resin supports.
Deleting a container now accepts a "purge" option that clears its /data.

@petrosagg @Page-

---
This change is [Reviewable](https://reviewable.io/reviews/resin-io/resin-supervisor/216)

### CODE REVIEW

File: src/utils.coffee [hunk 1]
```diff
@@ -204,3 +204,134 @@ exports.getOSVersion = (path) ->
...
+	'Labels'
+	'Cmd'
+	'Entrypoint'
+	'Image'
+	'Volumes'
+	'WorkingDir'
+	'NetworkDisabled'
+	'ExposedPorts'
+	'HostConfig'
+]
```
Comment (rchadalawada) [inline src/utils.coffee]: Need 'Name' also to be in this list. We use Name to name our containers in the create API

Comment (pcarranzav) [inline src/utils.coffee]: Done.

**Comment by Page-** [comment]
Reviewed 5 of 7 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/docker-utils.coffee, line 256 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJN2sR4V5btlOfnwOg:-KNJN2sR4V5btlOfnwOh:-1671547593) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L256)):*
> ```CoffeeScript
> 			.then (images) ->
> 				throw new Error('Only images created via the Supervisor can be used for creating containers.') if images.length == 0
> 				knex.transaction (trx) ->
> ```

It would be nice to use `tx` for the transaction, it's what we use in other places we handle transactions

---

*[src/docker-utils.coffee, line 283 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJMgqcnFIkPgkibqG5:-KNJMgqcnFIkPgkibqG6:1557762369) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L283)):*
> ```CoffeeScript
> 							utils.validateKeys(options.HostConfig, utils.validHostConfigOptions)
> 						.then ->
> 							docker.modem.dialAsync(optsf)
> ```

What's the reason for not using the standard dockerode create?

---

*[src/docker-utils.coffee, line 287 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJMXdDiwiqjD89fRuX:-KNJMXdDiwiqjD89fRuY:-408623240) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L287)):*
> ```CoffeeScript
> 							containerId = data.Id
> 							trx('container').update({ containerId }).where({ id })
> 							.then ->
> ```

This can be just `.return data`

---

*[src/utils.coffee, line 339 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJJBD8bm4yhJ4z3wO1:-KNJJBD9-2FaHclxONbH:-1694249237) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/utils.coffee#L339)):*
> ```CoffeeScript
> 		keys = _.keys(options)
> 		invalidKeys = []
> 		_.each keys, (key) ->
> ```

This can be replaced by
```
invalidKeys = _.without(keys, validSet...)
```
or
```
invalidKeys = _.keys(_.omit(options, validSet))

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*
<!-- Sent from Reviewable.io -->

**Comment by pcarranzav** [comment]
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/docker-utils.coffee, line 283 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJMgqcnFIkPgkibqG5:-KNKY1Cbwc83KkpSqEEY:-778478136) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L283)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> What's the reason for not using the standard dockerode create?
</details>
Cause we want to return the JSON data from the docker API, and dockerode returns a dockerode container object instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*

**Comment by petrosagg** [comment]
Review status: 5 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/docker-utils.coffee, line 256 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJN2sR4V5btlOfnwOg:-KNdQIsVV2LGVs_LPRZn:-502682548) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L256)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> It would be nice to use `tx` for the transaction, it's what we use in other places we handle transactions
</details>
@page- what is `tx`?

---

*[src/docker-utils.coffee, line 287 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJMXdDiwiqjD89fRuX:-KNdXaKch3eWx5111Gqf:132573536) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L287)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> This can be just `.return data`
</details>
We always put brackets on single line calls

---

*[src/utils.coffee, line 339 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJJBD8bm4yhJ4z3wO1:-KNdY0CtTs76y5eCO4Xl:160383625) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/utils.coffee#L339)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> This can be replaced by
> ```
> invalidKeys = _.without(keys, validSet...)
> ```
> or
> ```
> invalidKeys = _.keys(_.omit(options, validSet))
</details>
OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*

**Comment by Page-** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/docker-utils.coffee, line 256 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJN2sR4V5btlOfnwOg:-KNdisgTT3e-LeTESV9P:-194778705) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L256)):*
<details><summary><i>Previously, petrosagg (Petros Angelatos) wrote…</i></summary>
> @page- what is `tx`?
</details>
It's just a variable name we commonly use to stand for "transaction"

---

*[src/docker-utils.coffee, line 287 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJMXdDiwiqjD89fRuX:-KNdjL0SflNNdfgc6HY9:-1258708736) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L287)):*
<details><summary><i>Previously, petrosagg (Petros Angelatos) wrote…</i></summary>
> We always put brackets on single line calls
</details>
Always, except for the exceptions :P  I don't care either way in regards to brackets for this though

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*

**Comment by petrosagg** [comment]
Reviewed 2 of 7 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/docker-utils.coffee, line 256 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJN2sR4V5btlOfnwOg:-KNdjnTIk6h9o63wjIht:2010685241) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L256)):*
<details><summary><i>Previously, Page- wrote…</i></summary>
> It's just a variable name we commonly use to stand for "transaction"
</details>
is this comment resolved them? I'm a bit confused by revieable

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*

**Comment by Page-** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/docker-utils.coffee, line 256 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216#-KNJN2sR4V5btlOfnwOg:-KNdk71hkxiEP7IUmFl1:-2138039288) ([raw file](https://github.com/resin-io/resin-supervisor/blob/97971498594f4106fb3549f14f958115ef0a94ed/src/docker-utils.coffee#L256)):*
<details><summary><i>Previously, petrosagg (Petros Angelatos) wrote…</i></summary>
> is this comment resolved them? I'm a bit confused by revieable
</details>
It is by me, you'll see to the right of reply/acknowledge a little circle which expands to people involved in a conversation - you can use that to set yourself satisfied at any point (or use the acknowledge button, as long as you're not the last response)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*

**Comment by petrosagg** [comment]
Reviewed 3 of 7 files at r1, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/216)*

**Pull request merged**